### PR TITLE
gpui/zed: Allow setting a GPU preference

### PR DESCRIPTION
**Pull request by mrnugget**
This adds a new setting:

```jsonc
{
  "gpu": "discrete" // or: "integrated
}
```

On macOS, this setting will have an effect when selecting the device with which to render windows.

On Linux, it's without an effect.

See: https://github.com/zed-industries/zed/issues/5124#issuecomment-2227743811

Release Notes:

- Added a setting that allows users on macOS with Intel machines to define which GPU Zed should use: `{"gpu": "discrete"}`. The default is `"integrated"`, which is the lower-power GPU.

### CODE REVIEW

File: crates/gpui/src/platform.rs [hunk 1]
```diff
@@ -252,6 +252,16 @@ pub enum Decorations {
     },
 }
 
+/// A type to describe which GPU to prefer.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Default)]
+pub enum GPUPreference {
```
Comment (maxdeviant) [inline crates/gpui/src/platform.rs:257]: Can we follow Rust casing conventions here?

```suggestion
pub enum GpuPreference {
```

File: crates/zed/src/zed/gpu.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use settings::{Settings, SettingsSources};
+
+#[derive(Copy, Clone, Debug, Default, Serialize, PartialEq, Eq, Deserialize, JsonSchema)]
+#[serde(rename_all = "lowercase")]
+pub enum GPU {
```
Comment (maxdeviant) [inline crates/zed/src/zed/gpu.rs:7]: Similarly here:

```suggestion
pub enum Gpu {
```

File: crates/zed/src/zed/gpu.rs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        match self {
+            GPU::Discrete => gpui::GPUPreference::Discrete,
+            GPU::Integrated => gpui::GPUPreference::Integrated,
+        }
+    }
+}
+
+/// Settings related to the machine's GPU and how Zed uses it.
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, JsonSchema, PartialEq, Eq, Default)]
+pub struct GPUSettings {
```
Comment (maxdeviant) [inline crates/zed/src/zed/gpu.rs:24]:
```suggestion
pub struct GpuSettings {
```

**Comment by emilazy** [comment]
I wonder if a better approach might be to prefer external GPUs, but keep the preference of low‐power GPUs when deciding between built‐in GPUs? That would mean that people using eGPU setups would get the high performance they’ve already opted into by plugging in an external GPU, but Intel MacBook Pros would still get a good experience out of the box, without having to add an additional settings. It would be as simple as changing the sort key to `|device| (device.is_removable(), device.is_low_power())`.

On the other hand, if there are any Intel MacBook Pro users who really wish that the built‐in discrete GPU was being used, a setting would make more sense.

**Comment by emilazy** [comment]
I don’t mind if you open a PR :) One thing I’m unsure about is how it will behave when an external GPU is unplugged, but presumably any issues there would have already existed in Zed anyway, so it wouldn’t be a regression on the state before my first PR.

**Pull request closed without merging**